src/http.rs

```diff
@@ -0,0 +1,153 @@
+                        Ok(metadata) => metadata.len(),
+                        Err(_) => 0u64,
+                    };
+                    // if byte_count is zero don't pass range header
+                    match byte_count {
+                        0 => {
+                            client.get(url.clone())?
+                                .send()?
+                        }
+                        _ => {
+                            let byte_range = Range::Bytes(vec![ByteRangeSpec::AllFrom(byte_count)]);
+                            client.get(url.clone())?
+                                .header(byte_range)
+                                .send()?
+                        }
+                    }
+                } else {
+                    client.get(url.clone())?
+                        .send()?
+                }
+            }
+            None => {
+                client.get(url.clone())?
+                    .send()?
+            }
+        }
+
+        //client.get(url)?.send()?
+
+    } else {
+        client.get(url.clone())?
+            .send()?
+    };
+    print(&format!("HTTP request sent... {}",
+                   style(format!("{}", resp.status())).green()),
+          quiet_mode,
+          false);
+    if resp.status().is_success() {
+
+        let headers = if resume_download {
+            resp.headers().clone()
+        } else {
+            client.get(url.clone())?
+                .send()?
+                .headers()
+                .clone()
+        };
+        let ct_len = headers.get::<ContentLength>().map(|ct_len| **ct_len);
+
+        let ct_type = headers.get::<ContentType>().unwrap();
+
+        match ct_len {
+            Some(len) => {
+                print(&format!("Length: {} ({})",
+                               style(len).green(),
+                               style(format!("{}", HumanBytes(len))).red()),
+                      quiet_mode,
+                      false);
+            }
+            None => {
+                print(&format!("Length: {}", style("unknown").red()),
+                      quiet_mode,
+                      false);
+            }
+        }
+
+        print(&format!("Type: {}", style(ct_type).green()),
+              quiet_mode,
+              false);
+
+        print(&format!("Saving to: {}", style(fname).green()),
+              quiet_mode,
+              false);
+
+        let chunk_size = match ct_len {
+            Some(x) => x as usize / 99,
+            None => 1024usize, // default chunk size
+        };
+
+        let out_file = get_file_handle(fname, resume_download)?;
+        let mut writer = BufWriter::new(out_file);
+
+        let pbar = create_progress_bar(quiet_mode, fname, ct_len);
+
+        // if resuming download, update progress bar
+        if resume_download {
+            let metadata = fs::metadata(fname)?;
+            pbar.inc(metadata.len());
+        }
+
+        loop {
+            let mut buffer = vec![0; chunk_size];
```

Comment:
thinking a bit more, this might be annoying and/or impossible. going to think a bit more.

Reply:
I initialized the vector outside the loop and interestingly it takes more time downloading a file than when the vector is initialized inside the loop.

I'm not sure why, I have a hunch it could be when doing:
```rust
let bcount = resp.read(&mut buffer[..]).unwrap();
```
since we extract mutable slice of the entire vector and which after the first iteration will having some bytes.